apps/studio/src/server/modules/resource/resource.service.ts

```diff
@@ -231,3 +233,147 @@ export const moveResource = async (
     .where("id", "=", String(resourceId))
     .executeTakeFirstOrThrow()
 }
+
+// Returns a sparse IsomerSitemap object that revolves around the given
+// resourceId, which includes:
+// - The full path from root to the actual resource
+// - The immediate children of the resource (if any)
+// - The immediate siblings of the resource (if any)
+// - The immediate children of those siblings (if any)
+export const getLocalisedSitemap = async (
+  siteId: number,
+  resourceId?: number,
+) => {
+  if (resourceId === undefined) {
+    // Provide a default sitemap with a single resource
+    return {
+      id: "0",
+      layout: "content",
+      title: "Root",
+      summary: "",
+      lastModified: new Date().toISOString(),
+      permalink: "",
+    }
+  }
+
+  // Step 1: Get the actual resource
+  const resource = await getById(db, { resourceId, siteId })
+    .select(defaultResourceSelect)
+    .executeTakeFirstOrThrow()
+
+  // Step 2: Get all the ancestors of the resource
+  const ancestors = [resource]
+  let currentResource = resource
+
+  while (currentResource.parentId !== null) {
+    const parent = await getById(db, {
+      resourceId: Number(currentResource.parentId),
+      siteId,
+    })
+      .select(defaultResourceSelect)
+      .executeTakeFirstOrThrow()
+
+    ancestors.push(parent)
+    currentResource = parent
+  }
```

Comment:
just fyi, this is non locking so we are allowing the resource (as well as its parents) to change while this method is executing.

since this method is purely a `GET`, i actually dont see a problem but i thought this might be useful to point out

Reply:
also, i think a better way of doing this might be using recursive ctes  (see [here](https://www.postgresql.org/docs/current/queries-with.html))

Reply:
Good point, originally wanted to do it step-by-step but doing CTEs can also the steps understandable also. Swapped in bd051a9c8f9d1a82abde4ceb4c108498fb00a10a.